store/src/state.rs

```diff
@@ -342,27 +342,19 @@ impl State {
         account_ids: &[AccountId],
         note_tag_prefixes: &[u32],
         nullifier_prefixes: &[u32],
-    ) -> Result<(StateSyncUpdate, MmrDelta, MerklePath), anyhow::Error> {
+    ) -> Result<(StateSyncUpdate, MmrDelta), anyhow::Error> {
         let inner = self.inner.read().await;
 
         let state_sync = self
             .db
             .get_state_sync(block_num, account_ids, note_tag_prefixes, nullifier_prefixes)
             .await?;
 
-        let (delta, path) = {
-            let delta = inner
-                .chain_mmr
-                .get_delta(block_num as usize, state_sync.block_header.block_num as usize)?;
+        let delta = inner
+            .chain_mmr
+            .get_delta(block_num as usize, state_sync.block_header.block_num as usize)?;
```

Comment:
I think we are missing a `+1` here. This should be something like:
```Rust
let start_forest = block_num as usize + 1;
let end_forest =state_sync.block_header.block_num as usize; 
let delta = inner.chain_mmr.get_delta(start_forest, end_forest)?;
```

For example, let's say the MMR has 7 leaves. And the latest block known to the client is block 4 (i.e., `forest=5`). In the request, the client would use `block_num=4` (the number of the last block known to the client). But the delta they need is to go from 5-leaf MMR to 6-leaf MMR - that is `get_delta(5, 6)` (we don't go to forest 7 because we also send block header for block 6).

This will also require special handling for cases when `start_forest >= end_forest` (the case when `start_forest > end_forest` won't be organically handled by the `Mmr`, I believe). If this happens, we just send an empty delta.

We should also document this very well, both here and in the readme docs.

Reply:
I'm missing something here.

- `block_num` is 0-indexed
- `forest` is 1-indexed
- to convert from one to the other, we need `+1`

The above means we have to change to:

```rust
get_delta(block_num + 1, state_sync.block_header.block_num + 1)?;
```

- The Mmr `get_delta` is inclusive on both values.
- The request is non-inclusive, on the users's `block_num`, but inclusive on the `block_header.block_num`.

Meaning we have to increase the start by another `+1`:

```rust
get_delta(block_num + 2, state_sync.block_header.block_num + 1)?;
```

But, I guess @igamigo has tested this already, and it works? So maybe I missing something here?

Reply:
Humm, I guess it is because we are returning the `Mmr` delta to the previous block, because of the self-referential issue, so we also have a `-1` on both because of that.

Reply:
Yes, it is related to the self-referential issue but even without it, it could still make sense to have it work like this. Specifically:

1. For the `to_forest`, we probably still don't want to use `state_sync.block_header.block_num + 1` because we are sending the block header for `state_sync.block_header.block_num` to the client anyway and thus sending the delta for `forest = state_sync.block_header.block_num` makes sense.
2. For the `from_forest` we know that on the previous request the client had synced up to `forest = block_num + 1` (because we sent the delta up to `forest = block_num` + the block header for block at `block_num`), and so this is the value we should use for the starting forest on the new request.

Reply:
Added the `+1` and some docs.